### DISCUSSION TITLE
Is there any benchmark testing for segmenter wrt ICU4C WordBreakIterator?

### DISCUSSION THREAD

**Discussion by riajain0412** [unicode-org/icu4x#2948] [Q&A]
We currently use ICU4C WordBreakIterator for our use case. Recently, we came across ICU4X so thought of trying its wordBreakSegmenter. I analyzed the data size difference between ICU4C and ICU4X for WordBreakSegmenter for SEA languages and found that ICU4X's data size is slightly larger than ICU4C's data size and on the other hand your LSTM data size seems to be very less.
So I have below 2 queries:
1. As dictionary based segmentation data size is slightly large than our ICU4C data size so I want to know that how ICU4X's performance is better than ICU4C? Although the data size difference is in few KB's but we use it for WASM and even few KB's becomes a great deal for us because of WASM.
2. For LSTM based segmentation, I want to know it's accuracy details as well as how much is the latency if we use this model wrt to dictionary based segmentation?

**Comment by Manishearth** [comment]
ICU4X will likely be faster during data _loading_. However, note that our segmenter modules are experimental and we have not yet made the data fully zero copy, so there will be some situations where it's not as fast as it could be. When the segmenter modules are finished we will likely always be faster than ICU4C on data loading as we are in every other component.

For segmentation performance, I haven't measured it myself, but I think @zbraniecki has? I'll let the others give more specific answers as they have more context.

cc @makotokato @sffc @aethanyc

**Reply by riajain0412** [reply]
Okay, so what is the ETA for segmenter modules? And also how can we achieve lazyloading in ICU4X like loading files as and when needed?
@zbraniecki can you please help me in performance evaluation of segmenter wrt WordBreakIterator of ICU4C.

**Reply by Manishearth** [reply]
> Okay, so what is the ETA for segmenter modules? 

No, our segmenter modules _exist_ right now (you're already using them), they are just still experimental and need improvement.

I'm not sure if we have an ETA yet, I think we're hoping to have that this quarter.

But again, they already work, they should be fast, I'm just saying that they should get even better in the future.

> And also how can we achieve lazyloading in ICU4X like loading files as and when needed?

This strongly depends on how you know "when something is needed" but the provider_adapters crate should be helpful. We're working on better docs for this.

**Reply by zbraniecki** [reply]
@riajain0412 [here](https://docs.google.com/spreadsheets/d/1hmKWB4GkRbq9Fa66IrUCT6TNkIVywZtJ-l6Ah15ljW8/edit#gid=204550733) are the numbers we have based on [this](https://github.com/zbraniecki/icu-perf) harness.

**Reply by riajain0412** [reply]
Segmentation numbers are based on dictionary based segmentation right?

**Reply by riajain0412** [reply]
And what do you mean from ICU4X Blob(tailored)? Also, Can you please share the accuracy and performance details of LSTM segmenter?

**Reply by zbraniecki** [reply]
> Segmentation numbers are based on dictionary based segmentation right?

Yes, the test is *very* primitive. It's just the first ever test we conducted to establish a ballpark measurements. And in the land of i18n it's tricky because what's the "common" thing to test varies depending on your audience.
For now we tested just [a single german string](https://github.com/zbraniecki/icu-perf/blob/main/tests/segmenter/src/main.rs) (so, latin characters).

You're welcome to adjust the test to what you want to measure.

> And what do you mean from ICU4X Blob(tailored)?

It tests when we produce postcard data file with just the data needed by segmenter, vs. "all" of postcard for all components and locales (in which case the postcard file is very large of course).

> Can you please share the accuracy and performance details of LSTM segmenter?

I don't have accuracy or performance information for that. @makotokato or @aethanyc may have that. They're working on the Segmenter and aiming to replace the current one used in Firefox Layout engine with that.

**Reply by sffc** [reply]
You can read more about how LSTM works, the tradeoffs it makes, and its accuracy numbers versus ICU4C here:

https://github.com/unicode-org/lstm_word_segmentation

**Reply by sffc** [reply]
tl;dr, LSTM is quite accurate, even exceeding Dictionary accuracy in languages where we have good training data. It is slower than Dictionary; we have work in progress by @younies to improve that.

**Reply by riajain0412** [reply]
Thank you @sffc for sharing performance numbers of LSTM. Do you also know that how much it is(LSTM) slower than dictionary based segmentation?

**Reply by riajain0412** [reply]
I have one more query, where does CLDR data gets downloaded when we run icu4x-datagen command? And also can we create data files with icu_datagen with any data instead of CLDR's?

**Reply by sffc** [reply]
I don't have up-to-date numbers on the performance difference between dictionary and LSTM. But, even if I had microbenchmarks, they don't really tell the whole story about how the segmentation strategy affects the overall performance of a text layout engine in practice. If you segment one screen's worth of text at a time, it's unlikely that there will be a latency difference. If you are segmenting many pages of text, then segmentation becomes more critical to app latency and you may notice a difference.

I would actually be quite interested if you could take the performance measurements in your own context and share them.

**Reply by Manishearth** [reply]
Relatedly: what _is_ your context? What are you attempting to build with ICU4X?

**Reply by riajain0412** [reply]
> I don't have up-to-date numbers on the performance difference between dictionary and LSTM. But, even if I had microbenchmarks, they don't really tell the whole story about how the segmentation strategy affects the overall performance of a text layout engine in practice. If you segment one screen's worth of text at a time, it's unlikely that there will be a latency difference. If you are segmenting many pages of text, then segmentation becomes more critical to app latency and you may notice a difference.
> 
> I would actually be quite interested if you could take the performance measurements in your own context and share them.

Okay, Got it. Let me check what I can do to get performance measurements

**Reply by riajain0412** [reply]
> Relatedly: what _is_ your context? What are you attempting to build with ICU4X?

With ICU4X, I want to use WordBreakSegmenter for all the platforms(Win, Mac, android, wasm etc.) I'm currently using ICU4C for this so just wanted to analyze how is ICU4X better than ICU4C and what extra does ICU4X have to offer.

**Reply by Manishearth** [reply]
Right, but what are you building using the segmenter?

**Comment by zbraniecki** [comment]
> I analyzed the data size difference between ICU4C and ICU4X for WordBreakSegmenter for SEA languages and found that ICU4X's data size is slightly larger than ICU4C's data size 

@aethanyc @makotokato - do we understand the reason for ICU4X data to be slightly larger than ICU4C?
